packages/xo-server/src/api/vm.js

```diff
@@ -802,10 +801,24 @@ export const snapshot = defer(async function(
   }
 ) {
   const { user } = this
-  if (
-    vm.resourceSet !== undefined &&
-    (await this.getResourceSet(vm.resourceSet)).subjects.includes(this.user.id)
-  ) {
+  let resourceSet
+  try {
+    if (vm.resourceSet !== undefined) {
+      resourceSet = this.getResourceSet(vm.resourceSet)
+    }
+  } catch (error) {
+    if (noSuchObject.is(error)) {
+      console.warn(`Cannot find resource set ${vm.resourceSet}`)
+    } else {
+      throw error
+    }
+  }
+
+  if (resourceSet === undefined || !resourceSet.subjects.includes(user.id)) {
+    await checkPermissionOnSrs.call(this, vm)
+  }
```

Comment:
Why not move this in the `try` block?

Reply:
Why? If the resource set couldn't be found (`noSuchObject` error), we still want to check the permissions on the SRs to know if the user is allowed to snapshot the VM. Moving this in the `try` block would bypass the check in that case.